channeldb/db.go

```diff
@@ -1652,6 +1652,94 @@ func (c *ChannelStateDB) FetchHistoricalChannel(outPoint *wire.OutPoint) (
 	return channel, nil
 }
 
+func fetchFinalHtlcsBucket(tx kvdb.RTx,
+	chanID lnwire.ShortChannelID) (kvdb.RBucket, error) {
+
+	finalHtlcsBucket := tx.ReadBucket(finalHtlcsBucket)
+	if finalHtlcsBucket == nil {
+		return nil, kvdb.ErrBucketNotFound
```

Comment:
nit: this should be a defined error, something like `ErrNoChanDBExists`, otherwise we won't know which bucket is missing by looking at the error alone.

Reply:
The caller is wrapping it:

`return fmt.Errorf("cannot fetch final htlcs bucket: %w", err)`

Do you think that is insufficient?

Reply:
I see there is also a problem with the pr currently, because in `rpcserver.go`, I still match against `ErrHtlcUnknown`. Bubbling up those very granular errors also requires them to be matched higher up, which can be considered a layering violation.

Reply:
> Does every error need to contain an indication of the function that generated it? I am not sure if this is done generally.

Tbh I haven't thought about this until recently, with some frustrations while fixing itests. And my answer is no, simply because it can't be that simple😂 My experience is mostly from debugging the logs, and my hunch is that whenever we are crossing an abstraction we need to wrap the error, so we know its origins. That should be applied to errors returned from another package or an interface. 

And apparently, our code doesn't follow this practice, which is something I think we can improve, and that's my thought re wrapping errors.


> I still match against ErrHtlcUnknown. Bubbling up those very granular errors also requires them to be matched higher up, which can be considered a layering violation.

Totally agree with the layering violation. For this specific case, following what `fetchChanBucket` does is what I have in mind. Define something like `ErrNoFinalHtlcsBucketExists` and return it here. And to match it in rpcserver, we can `return fmt.Errorf("%w: fetch bucket: %v", ErrHtlcUnknown, err)` in `LookupFinalHtlc`.

Reply:
I see where you're coming from.

Returning specific errors like in `fetchChanBucket` is fine for me, but if I look for example here the error is plainly discarded:

https://github.com/lightningnetwork/lnd/blob/55d62f8bab7a9ceee530ec281c08f6d8b39fd53a/channeldb/channel.go#L2388

The specific wrapping that you are suggesting seem to be a new pattern in lnd. It looks like the inverse of how %w is commonly used. Don't you think that there is a limit to how much low-level error information you can carry across package boundaries? I'd think that logging may be a better way to deal with this to keep the layers separated cleanly.

Reply:
I did add specific errors for the two missing bucket cases, replacing `kvdb.ErrBucketNotFound`. Btw, the linter forcing `Error.Is` in switch statements doesn't feel completely right to me.

Reply:
> It looks like the inverse of how %w is commonly used. Don't you think that there is a limit to how much low-level error information you can carry across package boundaries? I'd think that logging may be a better way to keep the layers separated cleanly.

That's also true. I don't like the error chain becoming too long, and logging is definitely very helpful. But it seems like we don't do any loggings in `channel.go`. As for the new pattern, I think we need to gather more opinions. Also I'll try it out to see if it's really better in my other PRs.

> Btw, the linter forcing Error.Is in switch statements doesn't feel completely right to me.

I get it. But I just tell myself `errors.Is` is just like `==` with the ability to compare wrapped errors.

Reply:
My personal opinion is that `errors.Is` everywhere does not always make it better. Also for the common check `if ctx.Err() == context.Cancelled` it feels a bit weird. Almost like you don't really know what you're checking and instead use a catch-all to be sure. But ofc a minor nit in the grand scheme of things.

For this PR, shall I leave it as is now?

Reply:
Yeah I think it's good enough! Thanks for the work and discussion, it's been very helpful to me!

Reply:
Definitely not in love with `errorlint`: https://github.com/lightningnetwork/lnd/pull/7067/files#r1007218950